### ISSUE TITLE
Align OpenCRVS & MOSIP approach to addresses

### ISSUE THREAD

**Issue by euanmillar** [opencrvs/opencrvs-core#5874]
```mermaid
---
title: Approve registration correction as a Registrar
---
sequenceDiagram
    autonumber
    Client->>GraphQL gateway: approveRegistrationCorrection
    GraphQL gateway->>Workflow: POST /records/{recordId}/approve-correction
    Workflow->>Hearth: Save tasks (previous with accepted status and new REGISTERED status)
    Workflow->>Metrics: Notify metrics about the request
    Workflow->>Search: Request for bundle reindexing
    Workflow->>Notification: Send notification about approval to registration agent
    Workflow->>GraphQL gateway: Apply user submitted changes to record
    GraphQL gateway->>Hearth: Save bundle
```

**Comment by euanmillar** [comment]
Regarding changes in address, we prefer for MOSIP to expose a JSON API to us, therefore using Address Mapper with JSON, as opposed to Drools as a MOSIP backend for this API may be less work?  We can natively understand the Address Mapper with JSON.

Questions regarding mosip-context.json:

1. In the mosip-context json example,  why is dateOfBirth, bestTwoFingers, biometrics etc in the “address” container?

2. In the mosip-context json example, why is the new custom field “barangey” not listed to show how and at what level that custom area would be added?

3. In the mosip-context json example, why are location pincodes not represented in order to solve the _“potential for errors and data discrepancies.”_

**Comment by euanmillar** [comment]
I think that whatever the new JSON-LD standard we may create in DCI/G2PConnect, we need a way to go to and from FHIR (OpenCRVS) & OIDC(E-Signet) to the new standard. FHIR has no ability to extend strings like "city", and OIDC has no ability to extend strings like "locality" into JSON objects. Hence my request for a colon separator between readable label and id.

**Comment by rikukissa** [comment]
Just more in terms of the data structures selected here – is there a restriction in place preventing the use of something more flexible like an array structure?

```
[
   {
      type: "street_address",
      value: "1 Queen's Avenue"
   },
   { type: "locality", "value": { name: "Palawan", id: "56010000" },
   { type: "region", "value": { name: "Mimaropa", id: "56000000" },
   { type: "postal_code", "value": "postalCode" },
   { type: "country", "value": { name: "Philippines", code: "PHL" }
]
```

which could then be extended to support any administrative structure?

**Comment by naftis** [comment]
Think it's a good idea to let OpenID Connect return anything they like as an address, with a PSUT of the user. MOSIP then could have an endpoint where we could fetch the address details as JSON-LD format. Then we can also forget OpenID address format from the DCI response, and use the  JSON-LD address there as well. Then receiving end can decide if they want to convert it to OpenID connect or not.

**Comment by euanmillar** [comment]
I agree @naftis .  The implication is however, that essentially OpenCRVS ignores whatever OIDC address E-Signet sends us and instead uses a JSON-LD address from MOSIP.  Would that be OK from a consent / architectural point of view?

**Comment by naftis** [comment]
**Current problems**

1. We can have 1...n administrative levels. These levels are hardcoded all around the codebase and it's for example difficult to tell in backend that which level is the smallest
2. The FHIR Address has `state` and `district` fields, and `line[]` string array. Currently, Patient's address FHIR id's are stored in hardcoded indices in the line array, or in the `state` / `district` fields. We only need free text fields & a "partOf" field, which links to the smallest known administrative level for that location.
3. In some cases, the `state` / `district` fields have free text values instead of 

**Proposed solution**

1. Lets save the FHIR Address with an "part-of" extension like so. This means editing the FHIR builders and 
  
```json
{
  "resourceType": "Address",
  "line": ["123 Main Street", "Apt 1", "Floor 2"],
  "state": "Karnataka",
  "district": "Bangalore",
  "extension": [
    {
      "url": "http://opencrvs.org/specs/extension/part-of",
      "valueReference": {
        "reference": "Location/123"
      }
    }
  ]
}
```

2. Lets define the address forms with `dependsOn` like so:

```js
[{
  name: "state",
  type: "SELECT_WITH_DYNAMIC_OPTIONS",
  label: {...}
},
{
  name: "district",
  type: "SELECT_WITH_DYNAMIC_OPTIONS",
  label: {...},
  dependsOn: "state"
}]
```

3. Now FHIR builders can use the `dependsOn` field to determine which field is the smallest administrative level. For example, if the `state` field is the smallest administrative level, then the `state` field will be the last field in the address form. If the `district` field is the smallest administrative level, then the `district` field will be the last field in the address form.

4. Lets clean up the way we handle locations for example [here](https://github.com/opencrvs/opencrvs-core/blob/a3cffd46528fcb25dc63c08f9bb9f38ec6d202ae/packages/client/src/views/RecordAudit/utils.ts#L139), and lets rather resolve the `partOf` field upwards to get the full administrative structure.

**Comment by euanmillar** [comment]
Waiting for Mono to complete the work on the MOSIP side
